javascript/channel.ts

```diff
@@ -132,10 +142,15 @@ export class Channel {
         }
         if (msg.ID === codec.EofID) {
             this.gotEOF = true;
-            // if (this.readers.length > 0) {
-            // 	this.readers.shift()();
-            // }
-            return;
+            // TODO does this block if there are a lot of readers pending?
+            // is there a better way to interleave those?
+            while (true) {
+                let reader = this.readers.shift();
+                if (reader === undefined) {
+                    return;
+                }
+                reader();
+            }
```

Comment:
on EOF I needed to flush the pending readers so that they would consume whatever is left in the buffers

Reply:
there are some weird cases where it could. tight loops in js also lead to weird js implementation specific locking issues i've discovered (with this library, but more in the read loop not here)